Improve mat

akaviaLab opened the pull request:
* [X] fix #1040 
* [X] description of feature/fix
Improve reading of matlab models including annotations. See https://github.com/opencobra/cobratoolbox/blob/master/docs/source/notes/COBRAModelFields.md for fields added.
Fields I didn't add, and I'm not sure what they mean
all gene annotations (genes don't have annotations in cobrapy yet)
model.C, model.E. model.e*, model.D, model.d*

I haven't modified tests yet. Not sure what's necessary, but I'm thinking about making a copy of load_model which will load mat models from BIGG/BioModels, and then maybe make a test of load_model, load_model_mat and compare them.

I haven't modified the writing of mat models. If this code makes sense, I can write the inverse function.
* [X] add an entry to the [next release](../release-notes/next-release.md)
Improve mat reading

Commit:
speedup mat reading

Commit:
using dictionary comprehension to be a bit faster

Commit:
vectorized metabolite processing and added annotations

Commit:
vectorize reaction processing

Commit:
functionalize cell processing

Commit:
functionalize cell processing for floats

Commit:
added docstring

Commit:
added rxn annotations, including pubmed and EC

Commit:
added rxn annotations, including pubmed and EC

cdiener commented:
That's awesome, thanks so much!

I think all cobrapy species have annotations, including genes:

```
In [1]: from cobra.test import create_test_model

In [2]: mod = create_test_model("textbook")

In [3]: mod.genes[0].annotation
Out[3]: {}
```

Empty in this case but present.

Commit:
tidying

Commit:
functionalize annotations

cdiener commented:
With the [annotated Recon3D](https://www.vmh.life/files/reconstructions/Recon/3D.01/Recon3D_301.zip) it errors:

```
In [5]: rec = load_matlab_model("Recon3DModel_301.mat")
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 rec = load_matlab_model("Recon3DModel_301.mat")

File ~/code/cobrapy/src/cobra/io/mat.py:231, in load_matlab_model(infile_path, variable_name, inf)
    229 for possible_name in possible_names:
    230     try:
--> 231         return from_mat_struct(data[possible_name], model_id=possible_name, inf=inf)
    232     except ValueError:
    233         # TODO: use custom cobra exception to handle exception
    234         pass

File ~/code/cobrapy/src/cobra/io/mat.py:496, in from_mat_struct(mat_struct, model_id, inf)
    494         new_metabolite.formula = met_formulas[i]
    495     new_metabolites.append(new_metabolite)
--> 496 mat_annotations(new_metabolites, m, _d_replace=D_MET)
    497 model.add_metabolites(new_metabolites)
    499 new_genes = []

File ~/code/cobrapy/src/cobra/io/mat.py:368, in mat_annotations(target_list, mat_struct, _d_replace)
    364 annotation_matlab = list(
    365     set(mat_struct.dtype.names).intersection(D_REPLACE[_d_replace].keys())
    366 )
    367 annotation_lists = [[None]] * len(annotation_matlab)
--> 368 annotation_providers = [
    369     D_REPLACE[_d_replace + "_REV"][x] for x in annotation_matlab
    370 ]
    371 for i in range(len(annotation_matlab)):
    372     annotation_lists[i] = _cell_to_str_list(mat_struct[annotation_matlab[i]][0, 0])

File ~/code/cobrapy/src/cobra/io/mat.py:369, in <listcomp>(.0)
    364 annotation_matlab = list(
    365     set(mat_struct.dtype.names).intersection(D_REPLACE[_d_replace].keys())
    366 )
    367 annotation_lists = [[None]] * len(annotation_matlab)
    368 annotation_providers = [
--> 369     D_REPLACE[_d_replace + "_REV"][x] for x in annotation_matlab
    370 ]
    371 for i in range(len(annotation_matlab)):
    372     annotation_lists[i] = _cell_to_str_list(mat_struct[annotation_matlab[i]][0, 0])

KeyError: 'metKEGGID'
```

I think you don't want the `+ "_REV"` in line 369.

Commit:
fixed minor mistake with REV

akaviaLab commented:
Fixed.
If you can figure out what everything lower than 
model.E	m x evars 
on the cobratoolbox page means, and if it is relevant to cobrapy, I'd appreciate it. I think cobrapy doesn't support any of these fields, but if I'm wrong, please let me know.

cdiener commented:
Oh yeah, those are additional variables used for enzyme capacity models. Not sure whether we should read them. They can be represented with optlang, but cobrapy has no methods for enzyme capacity models at this point. We may want to support model.C, model.ctrs, and model.ctrNames, which would just be added as additional constraints to the model. But this should probably be its own PR. What we may consider in this PR is to add writing annotations to Matlab as well, or did you want this in a separate PR (would be fine with me)?

cdiener commented:
Some of the properties don't seem to get parsed correctly. For instance, EC numbers:

```
In [97]: m["Recon3DModel"]["rxnECNumbers"][0][0][12]
Out[97]: array([array(['1.4.3.6'], dtype='<U7')], dtype=object)

In [98]: m["Recon3DModel"]["rxns"][0][0][12]
Out[98]: array([array(['13DAMPPOX'], dtype='<U9')], dtype=object)

In [99]: rec.reactions[12].annotation
Out[99]: {'pubmed': '15035803,8920635,9653080'}

In [100]: rec.reactions[12]
Out[100]: <Reaction 13DAMPPOX at 0x7fb42ae47ca0>
```

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+        if annotation_matlab[i] == "rxnECNumbers":
+            # if there are more than one ec code, turn them to a comma separated str
+            # noinspection PyUnresolvedReferences,PyTypeChecker
+            annotation_lists[i] = [
```
Don't join them but keep them as lists. COBRAPY annoation value scan be lists for a single provider like `rx.annotation["ec-code"] = ["1.1.1.1", "1.2.1.1"]` for instance. Everything else won't be handled correctly by the SBML writer.

akaviaLab replied:
So single values should also be in a list?

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+        annotation_lists[i] = _cell_to_str_list(mat_struct[annotation_matlab[i]][0, 0])
+        if annotation_matlab[i] == "rxnReferences":
+            # noinspection PyUnresolvedReferences
+            annotation_lists[i] = [
```
Need to split them apart and save them as list as well.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+            ]
+        if annotation_matlab[i] == "metCHEBIID":
+            # if there are more than one ec code, turn them to a comma separated str
+            # noinspection PyUnresolvedReferences,PyTypeChecker
```
split them into a list

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+            annotation_lists[i] = [
+                x.replace("PMID:", "") if x else None for x in annotation_lists[i]
+            ]
+        if annotation_matlab[i] == "rxnECNumbers":
```
elif

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+                else None
+                for x in annotation_lists[i]
+            ]
+        if annotation_matlab[i] == "metCHEBIID":
```
elif

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+    )
+    annotation_lists = [[None]] * len(annotation_matlab)
+    annotation_providers = [D_REPLACE[_d_replace][x] for x in annotation_matlab]
+    for i in range(len(annotation_matlab)):
```
use more descriptive names for the running variable, and rather make annotation_lists a dict and iterate over the names like

```
annotations = dict.from_keys(annotation_providers, None)
....
for name in annotation_matlab:
....
```

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+        set(mat_struct.dtype.names).intersection(D_REPLACE[_d_replace].keys())
+    )
+    annotation_lists = [[None]] * len(annotation_matlab)
+    annotation_providers = [D_REPLACE[_d_replace][x] for x in annotation_matlab]
```
just calling it "providers" is probably sufficient.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+            annotation_providers[j]: annotation_lists[j][i]
+            for j in range(len(annotation_providers))
+            if annotation_lists[j][i]
+        }
```
Could be simplified to:
```
obj.annotation = {
    prov: annotation_lists[prov][i]
    for prov in annotation_providers
    if annotation_lists[prov][i]
}
```

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+            for j in range(len(annotation_providers))
+            if annotation_lists[j][i]
+        }
+        if len(annotation_dict):
```
Unnecessary, if the dict is empty you can still set it. The if probably generates more overhead.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -266,11 +432,14 @@ def from_mat_struct(
-        warn("Objective vector `c` not found.")
+    if "metCharge" in m.dtype.names and "metCharges" not in m.dtype.names:
+        warn(
+            "This model seems to have metCharge instead of metCharges field. Will use"
```
missing a space at the end of the first string part

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -284,76 +453,123 @@ def from_mat_struct(
-    coo = scipy_sparse.coo_matrix(m["S"][0, 0])
-    for i, j, v in zip(coo.row, coo.col, coo.data):
-        model.reactions[j].add_metabolites({model.metabolites[i]: v})
+    csc = scipy_sparse.csc_matrix(m["S"][0, 0])
```
We should try out what happens if we do this before running `model.add_reactions(new_reactions). There are some optimizations that kick in if you add reactions with assigned stoichiometry in bulk.

akaviaLab replied:
That is (with the other suggestions you made) much much faster

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -232,18 +342,74 @@ def create_mat_dict(model: Model) -> OrderedDict:
+            annotation_lists[i] = [
+                x.replace("CHEBI:", "") if x else None for x in annotation_lists[i]
+            ]
+    for i in range(len(target_list)):
```
Can use enumerate: `for i, obj in enumerate(target_list):`

Commit:
started annotations_to_mat

cdiener reviewed (changes requested):
Looking pretty good. Some improvements to variable names and simplifications.

akaviaLab reviewed (commented):
Okay. Will work on writing to matlab later.

akaviaLab commented:
> Oh yeah, those are additional variables used for enzyme capacity models. Not sure whether we should read them. They can be represented with optlang, but cobrapy has no methods for enzyme capacity models at this point. We may want to support model.C, model.ctrs, and model.ctrNames, which would just be added as additional constraints to the model. But this should probably be its own PR. What we may consider in this PR is to add writing annotations to Matlab as well, or did you want this in a separate PR (would be fine with me)?

If we can add these as constraints, you can tell me what each one means or how to represent it in optlang, so they can be read in. It can be this PR.
The enzyme capacity that we don't have methods in cobrapy should probably be ignored for now. 
As for writing annotations to matlab - I'm working on it.

Commit:
fast version of mat_annotations, need to split to list

Commit:
black and some tidying, including an else

Commit:
annotations will be lists

akaviaLab commented:
Some thoughts before mission creep and/or adding stuff to this PR that doesn't belong

I've been working on comparing RECON3D SBML and RECON3D mat reading to see what's missing, and I have some helper functions I'm working on and so forth.
1) Would you like me to try and add a function to compare models? I have a kludgy function I can tidy up. I'm using it to see where the differences are, so I can focus on what to fix. Separate PR is also an option.
2) Would you like me to improve SBML reading? For example, the RECON3D SBML (which I think was written by cobratoolbox) has things in Notes that should really be in annotations. I can add some changes to SBML notes reading/writing to make sure these kind of annotations end up in annotations, or I can ignore them.
3) It looks like cobratoolbox uses subsystems for groups, and only uses groups for reactions for now. Would you like me to make the matlab reader build groups based on subsystems? What about writing?

Commit:
fixed outputting of annotations. Added parsing of Rxn Notes from mat struct

Commit:
some minor name changes

Commit:
mat writing and reading work on all fields

Commit:
removed pymatbridge

cdiener commented:
> 1. Would you like me to try and add a function to compare models? I have a kludgy function I can tidy up. I'm using it to see where the differences are, so I can focus on what to fix. Separate PR is also an option.

I would say separate PR and maybe have it in the utils submodule?

> 2. Would you like me to improve SBML reading? For example, the RECON3D SBML (which I think was written by cobratoolbox) has things in Notes that should really be in annotations. I can add some changes to SBML notes reading/writing to make sure these kind of annotations end up in annotations, or I can ignore them.

We had that discussion before but Notes in SBML are specifically meant to not be parsed and are only there for a human to look at them. No machine readable information should be stored there. So ignoring those is fine. This should be fixed on the cobratoolbox side when creating those models.

> 3. It looks like cobratoolbox uses subsystems for groups, and only uses groups for reactions for now. Would you like me to make the matlab reader build groups based on subsystems? What about writing?

Yeah, the subsystem attribute itself is a leftover from legacy cobra models. Groups would definitely be the modern version for this. Probably better in a separate PR but it makes sense to read them as groups and write them from groups with specific names.

Commit:
fixed some errors

Commit:
black

cdiener commented:
Oh and single annotations don't have to be in a list right now, but they can be. And maybe leave the writing to a new PR because that one is already large.

Commit:
minor correction since number of arrays is inconsistent. Harvey and Harvetta should work now.

akaviaLab commented:
Oh and single annotations don't have to be in a list right now, but they can be. 

Can single annotations be in a list? I wrote the code in a way that both single and multiple annotations will go into a list.

The PR has writing. If you want me to separate it to a different PR, I could, but I did it before I saw your comment.

cdiener commented:
Yes a list is fine. You mean like `annotations = {"ec-number": ["1.2.3.4"]}`, right? That is fine.

And no need to separate the PR. I thought you hadn't started on writing yet, but it's all good.

akaviaLab commented:
Yes. That example is exactly how it is.

On Wed., Apr. 6, 2022, 11:08 p.m. Christian Diener, <
***@***.***> wrote:

> Yes a list is fine. You mean like annotations = {"ec-number": ["1.2.3.4"]},
> right? That is fine.
>
> And no need to separate the PR. I thought you hadn't started on writing
> yet, but it's all good.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobrapy/pull/1197#issuecomment-1091031309>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQYYZVUYALH3PFQIMZ7XSLVDZGSNANCNFSM5SGWEDZA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Commit:
removed unnecessary function and kludgy modification of MET_PROVIDERS

Commit:
modified regular expression. Added check for empty compartments which should not be happening

Commit:
ran black becase I had forgotten it

Commit:
added logger warning about compartments via regular expressions

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -681,6 +684,11 @@ def from_mat_struct(
         met_comps = [comps[i] for i in met_comp_index]
         met_comp_names = [comp_names[i] for i in met_comp_index]
     else:
+        logger.warning(
```
I think it would be cool if that warning also reports the names of all unique compartments. That way one can see immediately if there are wonky ones.

akaviaLab replied:
Like this new commit?

cdiener replied:
Yep, that looks good. I would drop the temporary assignment and just do `f"{', '.join(set(met_comps))}"`.

akaviaLab replied:
Done

Commit:
log compartments found

Commit:
some typing correction and test modification

akaviaLab commented:
Not sure which additional tests I should add without making the test/data directory very big, or more complicated functions.

Commit:
some more typing correction

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -218,6 +602,8 @@ def create_mat_dict(model: Model) -> OrderedDict:
     mat["rxns"] = _cell(rxns.list_attr("id"))
     mat["rxnNames"] = _cell(rxns.list_attr("name"))
+    annotations_to_mat(mat, rxns.list_attr("annotation"), D_REACTION_REV)
+    notes_to_mat(mat, rxns.list_attr("notes"), D_REACTION_NOTES_REV)
```
Notes in COBRA models should not be parsed. See the comments before. They also do not have to be a structured format and should generally be strings.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -179,6 +353,214 @@ def create_mat_metabolite_id(model: Model) -> str:
+            mat_dict[annotation_key] = _cell(_list)
+
+
+def notes_to_mat(
```
So notes from cobra model should never be parsed, and they should definitely never clash with, change, or modify annotations. They were used in legacy models for annotations, but this is not supported anymore with MIRIAM and other standards. So those should only be written as strings into the matlab metNotes or rxnNotes fields with some arbitrary formattting like several lines of "key: value" for instance. Int the future notes will probably become strings and can be written as is.

akaviaLab replied:
Okay.
This is again, not to lose information on round trips.
Something needs to happen with confidence scores, in order not to lose it.
The weird references go into notes from rxnReferences, and then they go back so they will be written in "rxnReferences". The whole _tmp_list2 section is for that.

cdiener replied:
I mean, it would already lose information that does not fit the identifiers that can be mapped to matlab. If you write them to the Matlab notes it would not get lost but put in a different place (and one that is consistent with cobrapy/SBML). But the point of the notes field is that it should not contain information that you want to keep or have parsed, that is what annotations are for.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -179,6 +353,214 @@ def create_mat_metabolite_id(model: Model) -> str:
+    )
+    note_providers = [D_REPLACE[_d_replace][x] for x in annotation_matlab]
+    notes = dict.fromkeys(note_providers, None)
+    _pumbed_re = re.compile("PMID: ?(\\d+),?")
```
typo, should be `_pubmed_re`. Also, it's already a variable limited to the function scope so the `_` prefix is not required.

akaviaLab replied:
Typo corrected. I feel better keeping it starting with _, since there are multiple variants of it. Each one is limited in scope, but that tells me it should be limited when reading the code.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -179,6 +353,214 @@ def create_mat_metabolite_id(model: Model) -> str:
+        }
+
+
+def mat_parse_notes(
```
Not really required but nice to have I guess. Though I would argue, they should still go into the annotation field and not into notes. Notes should be taken from the matlab metNotes and rxnNotes fields only. This would work with the SBML writer as well which may issue warnings that the annotations have no MIRIAM identifier but will not crash.

akaviaLab replied:
Clarifying.
There are situations where the rxnReferences in matlab has something like
'PMID:9028. Also when we read this book by Blah et al, 2001, published Random House' This can't be an identifier, so I'm not moving it into annotations, but notes. I don't want to throw it away, since it is info

The mat_to_notes function takes the notes from three fields (rxnReferences as above, rxnConfidenceScores, rxnNotes), throws away anything that can go into annotations (that is in identifiers), and keeps the rest.
It is in a dictionary, because Confidence Scores are still kept as a distinct entity, even if COBRA objects dont' have it. Otherwise, we will lose data on round trip back and forth to mat, and some people care about Confidence Scores. 
SBML reader still reads notes into a dictionary as well.
It could be a string, but then read-write matlab would loose information, including confidence scores which people care about.

cdiener replied:
Oh god, yeah I understand now. I guess it makes sense to have those in the notes then. But I would leave them there. The issue with confidence scores saved outside of notes or MIRIAM annotations is that there is no way to interpret them. For instance, what do confidences of 100, 0, and -10 mean if they appear? Is a confidence of 3 in model A the same as in model B? So in the end, that information is hard to use outside a standardized system.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -284,139 +673,148 @@ def from_mat_struct(
+        }
+        new_reactions[i].add_metabolites(stoic_dict)
 
-def model_to_pymatbridge(
```
Good for me to deactivate that since we have been looking for someone to test this for 2+ years now and nobody has stepped up. We should keep in mind that this requires a new release though since it removes existing functionality.

akaviaLab replied:
Okay

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -284,139 +673,148 @@ def from_mat_struct(
+        coefficients = dict(zip(new_reactions, c_vec))
+        set_objective(model, coefficients)
+    else:
+        warn("Objective vector `c` not found.")
```
I would transfer all wanrings to logger warnings since those can be manager better and play nice with multiprocessing (thread safe).

akaviaLab replied:
Done

cdiener reviewed (changes requested):
Looks fantastic. Some changes to how notes are handled to make it more future-proof. Regarding the tests we may want to look for a small Matlab model that has as many features/annotations as possible. I suspect the low coverage is due to many annotations not being present in the models, so those parser sections are not tested. Or we could try to take one of the well-annotated SBML models like textbook and write it, then read it...

Commit:
some corrections according to comments

akaviaLab reviewed (commented):
Corrected where I could.
SBML still reads NOTES into dictionary, which is partially necessary because there are notes like "Confidence Scores" that people want to keep distinct. I don't care, but I think people still do (see #653). And I think the cobratoolbox still uses confidence scores.
If there is a global decision, it should be global and apply to SBML and all other formats as well.

cdiener commented:
> SBML still reads NOTES into dictionary, which is partially necessary because there are notes like "Confidence Scores" that people want to keep distinct. I don't care, but I think people still do (see https://github.com/opencobra/cobrapy/issues/653). And I think the cobratoolbox still uses confidence scores.
If there is a global decision, it should be global and apply to SBML and all other formats as well.

This has more to do with cobrapy default types and incomplete support of SBML annotation namespaces. SBML already allows for custom annotations, but we don't support that yet. I think at some point, there were also plans for key-value types in FBC that would take this function. Notes in SBML are just strings with no formatting requirements, and we also don't support that  well yet. But we probably will at some point, and then all those instances that assume or require that notes are a dict will have to be adjusted.

akaviaLab commented:
Lots of comments, so I'm trying to have one reply. I'm writing what I understand. Feel free to correct me.

1) Should rxnNotes go into cobra.notes? I think the answer is yes.
2) Should cobra.notes go out into rxnNotes? Yes, because otherwise reading/writing a mat file will lose info. 
Missing annotation keys can be added later. I covered the ones in the cobratoolbox git page. If someone says "I have my annotation which isn't read" we can add the key.
3) Are notes still a dictionary? Yes, for now. If not, what should they be?
4) What to do with mat fields that absolutely do not match the identifiers format?
Like confidence scores or full references that do not match pubmed at all?
If they need to go in annotation instead of notes, I would give them an annotation key that starts with _ (which signifies won't work with identifiers)
So it could be something like
annotation = {
  '_confidenceScores' = 
  '_rxnReferences' = 
}
And then they can be written out to mat files and ignored otherwise.

If they are going into notes, what is the format (question 3).

cdiener commented:
>  1. Should rxnNotes go into cobra.notes? I think the answer is yes.

Yep, should go into `Reaction.notes`. Since they are currently a dictionary you could choose a key. May make sense to keep the source around, so it could be `rxn.notes["cobratoolbox"] = rxnNotes[i]`.

>  2. Should cobra.notes go out into rxnNotes? Yes, because otherwise reading/writing a mat file will lose info.

Yes, exactly.

>  Missing annotation keys can be added later. I covered the ones in the cobratoolbox git page. If someone says "I have my annotation which isn't read" we can add the key.

You mean reading from cobratoolbox? I think what you covered here is pretty complete. I mean writing COBRAPY notes to matlab, but this would be covered by (2) if we go that direction.

>  3. Are notes still a dictionary? Yes, for now. If not, what should they be?

Yes, see above for one idea. But any key should be fine.

> 4. What to do with mat fields that absolutely do not match the identifiers format?
> Like confidence scores or full references that do not match pubmed at all?
> If they need to go in annotation instead of notes, I would give them an annotation 
> key that starts with _ (which signifies won't work with identifiers)
> So it could be something like ...

I think it's okay to write them to notes, since they will need human intervention to be interpreted. That means giving up a perfect round trip, but that is fine. If you put them into annotations they **will be lost when saving to SBML**. So if you want them to be conserved you can keep them in the notes for now. 

> If they are going into notes, what is the format (question 3).

Using the matlab identifiers is fine, like `rxn.notes["confidence"] = ..., etc`. The only thing where I suggested a different route is writing them into matlab, where I would just save everything in the `rxnNotes` even if that means that confidence scores change their location in the matlab model to make it easier on us. If notes ever get converted to string they would enter the notes as `<p>confidence: 3</p>\n...`, but hopefully we would be able to write them to an actual key-value store at this point that is supported by SBML and our IO.

---

I am open for other ideas though. Let me know what you think.

akaviaLab commented:
Okay.
If notes are still a dictionary, it isn't more complicated to write them to matlab fields, and less disappearing data.

Confidence Scores are put into a key called 'Confidence Level' when reading from SBML, so that's why it is identical. So mat -> cobra -> sbml will write them in a format that can be read by sbml. cobratoolbox already has a specific field, so there's no point in putting them in rxnNotes and leaving rxnConfidenceScores empty or missing.

I can put rxnReferences that aren't MIRIAM compliant into cobrapy.notes, and then write them into cobratoolbox.rxnNotes.
cobratoolbox.rxnNotes -> cobrapy.notes["NOTES"] but it could be cobrapy.notes["rxnNotes"]

I'll add a comment that when SBML supports note strings, cobrapy should be revised, including mat reader/writer.

akaviaLab commented:
Changes should hopefully answer this.
For tests, thinking about making a function to download MAT models from BIGG, and then a test to download MAT and SBML, and compare them. If that sounds like a good idea, is there a way to convert BytesIO to open File-like object? I'm asking since scipy_io.loadmat needs an str or open file-like handle. If you don't think load_matlab_web() function should be in this PR, I'll find a different test.

Commit:
some corrections according to discussion on PR

cdiener commented:
In general you can use the pytest tempfile fixture to write it out then read it. BiGG wouldn't work though since the mat models are actually generated with cobrapy so they miss all the annotations not supported before. Also we want the tests to be fast if possible. So the smallest model with the most annotations we can find.

akaviaLab commented:
Okay then.
Maybe I'll add compressed recon3.mat if it isn't too big.

On Mon., Apr. 11, 2022, 7:44 p.m. Christian Diener, <
***@***.***> wrote:

> In general you can use the pytest tempfile fixture to write it out then
> read it. BiGG wouldn't work though since the mat models are actually
> generated with cobrapy so they miss all the annotations not supported
> before.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobrapy/pull/1197#issuecomment-1095700216>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQYYZUVH3EGQ7C6RQPS4X3VES2MDANCNFSM5SGWEDZA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Commit:
some corrections according to discussion on PR and cobratoolbox test models

Commit:
some corrections according to cobratoolbox test models

Commit:
fix bug in GPR.from_string()

Commit:
updated tests. Was treating logger.warning like warn.warning but the function is different. Also, the documentation states that
"warnings.warn() in library code if the issue is avoidable and the client application should be modified to eliminate the warning"
And malformed GPR should be modified in the client application (by modifying the input).

Commit:
added subsystem to groups. some tidying

Commit:
corrected according to comments.

Commit:
removed deprecated function remove_from_model in gene.py

Commit:
removed unnecessary functions now need to update tests for coverage

Commit:
not using enumerate

akaviaLab commented:
Adding test cases, and nearly done.
Question - models that don't have compartment in the met_id (like raven.mat) that match with 
if all(var in m.dtype.names for var in ["metComps", "comps", "compNames"]):

Are rewritten (because of create_mat_metabolite_id) as metabolites that do have compartment in the met id.
For example, in raven.mat
'm156' is rewritten as 'm156[Cytoplasm]'

It means that models that don't have compartments in met_ids do not make a seamless read, write, reread transfer.
Should this be happening? Should I change raven.mat or change the functions so it is rewritten the same as it was?

This seems to be a quirk of RAVEN format, so I don't know if it is even used apart from test cases.

Commit:
added some tests

cdiener commented:
Yeah, I agree with you that this is odd. I would also favor maintaining IDs the same. There isn't really a rule that compartment has to be in a metabolite ID. It's a convention that many  IDs follow, but you usually also want a real attribute for a compartment. I would vote for removing that substitution and using the original ID. My argument would be that it would be surprising if I wrote a COBRApy model to Matlab and suddenly all my metabolites got renamed.

Commit:
raven style model should be written properly
made failure better
added tests to fail when compartments are missing or problematic
added test to see that mat models with no genes are imported correctly

akaviaLab commented:
It looks like pepole who create mat files are sloppy and/or the behavior of the cobratoolbox changed over time. Specifically, the field names for annotations keep varying a bit on the capitalization
metChEBIID, metCHEBIID, and metChebiID are things that I've seen

Likewise for other fields.

Question - do we want cobrapy to be tolerant of this? If not, I'll go with the official description in https://github.com/opencobra/cobratoolbox/blob/master/docs/source/notes/COBRAModelFields.md

If so, do you have any idea on how to do case-insensitive matching?
Right now, I run 
annotation_matlab = list(
        set(mat_struct.dtype.names).intersection(D_REPLACE[_d_replace].keys())
    )
Any ideas on how to do case-insensitive matching equivalent to this? I could do regular expression, but that seems like it would be complicated.

cdiener commented:
Yeah, they run no validations at all it seems. You could convert all keys in the mat struct to lower case and match on those only, maybe?

akaviaLab commented:
Okay. I'll output what is supposed to be the official version.

Commit:
cobratoolbox created models seem to be inconsistent about capitalization of model fields (metChebiID, metCHEBIID, metChEBIID) - revised reading to be tolerant of this

Commit:
CHEBI: shouldn't be removed from metCHEBIID since that how it is accessed in the CHEBI website.
removed unnecessary function create_mat_metabolite_id()

Commit:
mat models seem to be read and rewritten correctly

Commit:
read gene names and make sure to add genes to model (otherwise gene names are ignored)

Commit:
fix for old cobratoolbox fields
added tests - test_compare_xml_to_written_matlab_model, test_mat_model_with_long_compartment_ids, test_mat_model_wrong_caps

Commit:
save model name only if it exists

akaviaLab committed:
Merge branch 'devel' into improve_mat

akaviaLab commented:
All done. Including tests for all kinds of weird cases.

cdiener commented on src/cobra/test/test_io/test_mat.py:
```diff
@@ -68,10 +69,257 @@ def test_save_matlab_model(
+    assert compare_models(raven_mat_model, raven_mat_model_reload) is None
+
+
+def _fix_xml_annotation_to_identifiers(model: "Model") -> None:
```
That was a great catch. The annotations in those SBML models are indeed wrong. Though, rather than fix it in the test ad hoc we should probably fix those in a different PR in the models themselves.

akaviaLab replied:
I can do that, but then the test would fail. Remove it temporarily and return it when the models are improved?

cdiener replied:
Leave it in and remove it in the next PR.

akaviaLab replied:
The comparison function doesn't pick up on the differences (I was using a more strict one for development), so I can remove the fixing function. The later PR will fix the models to work even with strict comparison.
OR I can just stop changing this PR and do it all in a later one.

cdiener commented on src/cobra/test/test_io/test_mat.py:
```diff
@@ -68,10 +69,257 @@ def test_save_matlab_model(
+    for gene in model.genes:
+        if len(gene.annotation):
+            if "ncbigi" in gene.annotation.keys():
+                gene.annotation["ncbiprotein"] = gene.annotation.pop("ncbigi")
```
I thinks this should be `gene.annotation["ncbigene"] = ...`

akaviaLab replied:
Well, the model has ncbigi, not ncbigene, which is an identifier in Genbank (abbreviated as GI) for proteins. It wouldn't make sense for it to be ncbigene.

cdiener commented on src/cobra/test/test_io/test_mat.py:
```diff
@@ -68,10 +69,257 @@ def test_save_matlab_model(
+        )
+
+
+"""
```
What is this supposed to do? Is this just there to generate the model for the tests? In that case you can remove it here. You can create a Gist if you want and link to it in a comment.

akaviaLab replied:
Done

cdiener commented on src/cobra/test/test_io/test_sbml.py:
```diff
@@ -205,7 +205,9 @@ def test_read_2(self, io_trial: Tuple[str, Model, Model, Model]) -> None:
         """
         name, reference_model, test_model, _ = io_trial
-        if name in ["fbc1", "mat", "cobra", "raven-mat"]:
+        if name in [
```
Why are you changing SBML tests here?

akaviaLab replied:
Not sure why I was doing that. It doesn't fail, but I can change it back.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -71,6 +179,73 @@ def _cell(x: Iterable[str]) -> np.ndarray:
+def _cell_to_str_list(
+    m_cell: np.ndarray,
+    empty_value: Optional[str] = None,
+    _re: Optional[Pattern] = None,
```
Don't use underscore naming in function arguments they are local anyway. Rather try to give them a clear and interpretable name.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -158,25 +334,239 @@ def save_matlab_model(
+    annotation_matlab = list(set(struct_names).intersection(matlab_field_dict.keys()))
+    providers = [matlab_field_dict[x] for x in annotation_matlab]
+    annotations = dict.fromkeys(providers, None)
+    _pubmed_re = re.compile("PMID: ?(\\d+),?")
```
i think those are constants and can go on top with the other precompiled regexes.

akaviaLab replied:
Done

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -158,25 +334,239 @@ def save_matlab_model(
+    matlab_field_dict = {
+        x.casefold(): D_REPLACE[_d_replace][x] for x in D_REPLACE[_d_replace].keys()
+    }
+    _caseunfold = {x.casefold(): x for x in mat_struct.dtype.names}
```
No need to use underscore for local variables. They are not exported anyways.

cdiener commented on src/cobra/io/mat.py:
```diff
@@ -158,25 +334,239 @@ def save_matlab_model(
+def mat_parse_notes(
+    target_list: List[Object],
+    mat_struct: np.ndarray,
+    _d_replace: str = D_REACTION_NOTES,
```
Drope the underscore. Also what does the "D_" stand for? Maybe improve the naming.

akaviaLab replied:
Dropped and improved. The D_ was for DICT_ so that's what it is now.

cdiener commented on release-notes/next-release.md:
```diff
@@ -15,6 +18,8 @@ fail more comprehensibly.
 
 ## Other
 
+Removed pymatlib direct transfer of models to matlab process. 
```
This has to go in backwards incompatible changes.

cdiener reviewed (changes requested):
Okay, only a bunch of formatting/variable naming changes. The code is getting really hard to review at this point due to large number of commits. So I would probably recommend to leave most of those for a clean-up PR afterwards. The only thing for now would be to remove the code in the docstring that generates the wrong key model. Then we can merge and follow up with a PR to fix the wrong annotations in SBML models and another one for the variable name cleanup.

akaviaLab reviewed (commented):
Fixed. Some questions remaining.

Commit:
variable name fixes
next release corrected
removed instructions to gist

Commit:
revert changes in test_sbml.py

akaviaLab commented:
@cdiener - reverted the test_sbml.py changes
Will change the models and everything in a later PR. I'll add a note to the new PR to be merged after this one.
Yuck - this is not just changing the XML files, but changing how they are read. I'll open a discussion for that.

cdiener commented:
Actually, my bad. I read up on it and I think the lower case providers in SBML are now the de facto standard and are allowed by identifiers.org. So annotations["sbo"] is fine and the SBML parser will consolidate into this format (so transform from "SBO"). But we can discuss this more in another PR that addresses this because it also needs some adaptations for IDs.

akaviaLab commented:
@cdiener - I think we're done. Would you like another developer to review this?
I have the various improvements and cleanup ready to go in another PR once this is merged.

cdiener reviewed (approved):
Let's get this in.
